core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala

```diff
@@ -308,6 +309,14 @@ class HadoopRDD[K, V](
     // Do nothing. Hadoop RDD should not be checkpointed.
   }
 
+  override def persist(storageLevel: StorageLevel): this.type = {
+    if (storageLevel.deserialized) {
+      throw new SparkException("Can't cache HadoopRDDs as deserialized objects because Hadoop's" +
```

Comment:
+1; I had a recent PR that added some similar errors / warnings for common user mistakes, but I only raised exceptions for branches of the code that threw errors (e.g. NPE) 100% of the time.  Let's make this into a warning.